### ISSUE TITLE
Layer interrogation

### ISSUE THREAD

**Issue by qinghezeng** [mahmoodlab/CLAM#87]
Hi, thank you for your excellent work! 

It would be my pleasure to cite it in my research. I wonder if you have thought about the layer interrogation? Given that the ResNet50 was trained on ImageNet without fine-tuning, do you think it make sense to interrogate its layers? Is there any technique we can apply especially for the MIL part? Thank you again!

**Comment by fedshyvana** [comment]
hello, I am not familiar with what you mean by "layer interrogation" actually, any examples/materials that you might reference?

for MIL, one thing that i didn't do in the paper but later found to help was to increase the depth of the MIL network (e.g. add more hidden layers, increase width of attention network, etc.) and it might also be helpful to apply Dropout to the actual attention scores (instead of just hidden layers) aka zeroing out attention scores of e.g. 20% of patches when performing attention pooling. Intuitively this might allow the model to learn to attend to a more diverse panel of morphologies (if they exist in your data) otherwise, the model might learn to only assign high attention to a single type of morphology and ignore others or try multi-headed attention similar to what is used in transformers. 

Additionally, feel free to experiment with 1. different feature representations (pretrained ResNet vs. self-supervised embeddings) and 2. different learning algorithms: GNN vs. MIL vs. VIT, etc.

**Comment by qinghezeng** [comment]
Hi, apologizes for my late reply! @fedshyvana 

Sorry for not being clear. Actually I meant layer-wise interpretation. For example the following reference:
https://arxiv.org/pdf/1610.01644.pdf
https://arxiv.org/pdf/1904.04520.pdf
https://github.com/albermax/innvestigate
https://github.com/lmcinnes/umap
https://github.com/pytorch/captum

Most of them are based on Tensorflow but I found it not so easy to convert the CLAM from Pytorch to Tensorflow (for example using `onnx` will failed with the `transpose` operation). 

And I also wonder if you think it make sense to interprete the layers of ResNet50 since it is trained on ImageNet. Moreover, it seems that there is no layer-wise interpretation designed for MIL, any idea? Thanks a lot in advance!!

Many thanks to the nice suggestions about MIL and the other experiments! They are most likely perform better.
